### PR TITLE
Collin/workflow aliases

### PR DESCRIPTION
**Pull request by cwebbles**
Add ability for workflows to register themselves under an alias with the workflow manager.

### CODE REVIEW

File: quest_test/test_alias.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+# TODO: test exception on alias dict collision
+
+@pytest.mark.asyncio
+async def test_alias():
+    pause = asyncio.Event()
+    data_a = []
+    data_b = []
+
+    async def workflow_a():
```
Comment (cwebbles) [inline quest_test/test_alias.py:20]: Not sure how the resources work, if I should access the queue within the alias or outside.

File: quest_test/test_alias.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        if wid not in histories:
+            histories[wid] = PersistentHistory(wid, InMemoryBlobStorage())
+        return histories[wid]
+
+    async with WorkflowManager('test_alias', storage, create_history, lambda w_type: workflow_a) as manager:
+        manager.start_workflow('workflow_a', 'wid_a')
+
+        await asyncio.sleep(0.1)
+
+    # async with WorkflowManager('test_alias', storage, create_history, lambda w_type: create_workflow) as manager:
```
Comment (cwebbles) [inline quest_test/test_alias.py:59]: Scenario with two workflows trading with the same alias

File: src/quest/manager.py [hunk 3]
```diff
@@ -151,3 +165,26 @@ async def get_event(self, workflow_id: str, name: str, identity: str | None) ->
...
+    # TODO refactor, deduplicate
+    def _check_alias(self, alias: str) -> str | None:
+        """Check to see if an alias exists
+
+        Returns a str of the workflow id if alias exists, None if it doesn't
+        """
+        return self._alias_dictionary.get(alias)
+
+def find_workflow_manager() -> WorkflowManager:
+        if (manager := workflow_manager.get()) is not None:
```
Comment (cwebbles) [inline src/quest/manager.py:189]: Having the same issue here we saw in class. `workflow_manager` exists, but `.get()` returns a LookupError

File: quest_test/test_alias.py (hunk 1 above)
Comment (byubean) [reply quest_test/test_alias.py:20]: Open the queue first. That way it will stay open for longer than the alias is in scope.

Comment (byubean) [reply quest_test/test_alias.py:20]:
```py
async with queue('data', None) as q:
    async with alias('the_foo'):
        data_a.append(await q.get())
    await pause.wait()
    data_a.append(await q.get())
```

File: quest_test/test_alias.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    #     await asyncio.sleep(0.1)
+    #     # await manager.wait_for_completion('wid_a', None)
+    #     # await manager.wait_for_completion('wid_b', None)
+
+    async with WorkflowManager('test_alias', storage, create_history, lambda w_type: create_workflow) as manager:
+        # Gather resources
+        # TODO: Do I need to grab this again after the alias switches?
+        foo_resources = await manager.get_resources('the_foo', None)
+        a_resources = await manager.get_resources('wid_a', None)
+        b_resources = await manager.get_resources('wid_b', None)
```
Comment (byubean) [inline quest_test/test_alias.py:74]: For this test, I would use the `send_event` function directly:
```py
await manager.send_event('wid_a', 'data', None, 'put', 'data a 1')
await manager.send_event('wid_b', 'data', None, 'put', 'data b 1')
await manager.send_event('the_foo', 'data', None, 'put', 'data foo 1')
```

we may need to change the workflow to something like:
```py
async def workflow_a():
    async with queue('data', None) as q:
        async with alias('the_foo'):
            await first_pause.wait()
            data_a.append(await q.get())
            data_a.append(await q.get())
        await second_pause.wait()
        while q:  # do we have the ability to see if the queue has items?
            data_a.append(await q.get())
```

Do the same for workflow b.

This way we send data to "wid_a", "wid_b", and "the_foo" twice. The first time two items go to workflow A. The second time two items go to workflow B.

At the end, simply assert the full contents of `data_a` and `data_b`:
```py
assert data_a == ['data a 1', 'data foo 1', 'data a 2']
assert data_b == ['data b 1', 'data b 2', 'data foo 2']
```

Comment (byubean) [reply quest_test/test_alias.py:74]: The first pause makes sure the queues are set up before we try to push to them.
The second pause gives the workflows a chance to swap the alias. 

```py
async def workflow_b():
    async with queue('data', None) as q:
            await first_pause.wait()
            data_a.append(await q.get())
        await second_pause.wait()
        async with alias('the_foo'):
            await third_pause.wait()
            data_a.append(await q.get())
            data_a.append(await q.get())
```

Comment (byubean) [reply quest_test/test_alias.py:74]:
```py
manager.start('wid_a'...)
manager.start('wid_b'...)
first_pause.set()
await manager.send_event('wid_a', 'data', None, 'put', 'data a 1')
await manager.send_event('wid_b', 'data', None, 'put', 'data b 1')
await manager.send_event('the_foo', 'data', None, 'put', 'data foo 1')
await asyncio.sleep(0.1)  # yield to the workflows

# now both should be waiting on second gate and no one should be the foo
assert not manager.has_workflow('the_foo')
second_pause.set()
await asyncio.sleep(0.1)  # yield

# now workflow b should be the foo
await manager.send_event('wid_a', 'data', None, 'put', 'data a 2')
await manager.send_event('wid_b', 'data', None, 'put', 'data b 2')
await manager.send_event('the_foo', 'data', None, 'put', 'data foo 2')
third_pause.set()
await asyncio.sleep(0.1)  # yield to the workflows

assertions
```

File: quest_test/test_alias.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+@pytest.mark.asyncio
+@timeout(3)
+async def test_alias():
+    first_pause = asyncio.Event()
+    second_pause = asyncio.Event()
+    data = []
+
+    async def workflow():
+        async with queue('data', None) as q:
```
Comment (cwebbles) [inline quest_test/test_alias.py:20]: This queue doesn't seem to get properly initialized due to a `find_historian()` failing with a `LookupError` in `InternalResource.__init__()`

File: src/quest/manager.py [hunk 6]
```diff
@@ -151,3 +166,25 @@ async def get_event(self, workflow_id: str, name: str, identity: str | None) ->
     async def get_identity_queue(self, workflow_id: str, name: str, identity: str | None) -> IdentityQueue:
         await self._check_resource(workflow_id, name, identity)
         return self._wrap(IdentityQueue(), workflow_id, name, identity)
+
+    async def register_alias(self, alias: str, workflow_id: str):
```
Comment (byubean) [inline src/quest/manager.py:170]: Let's make these private: `_register_alias` and `_deregister_alias`

We don't want developers calling these directly.

File: src/quest/manager.py [hunk 7]
```diff
@@ -52,7 +59,9 @@ async def __aexit__(self, exc_type, exc_val, exc_tb):
             await historian.suspend()
 
     def _get_workflow(self, workflow_id: str):
-        return self._workflows[workflow_id]  # TODO check for key, throw error
+        if (alias := self._check_alias(workflow_id)) is not None:
+            return self._workflows[alias]
+        return self._workflows[workflow_id]
```
Comment (byubean) [inline src/quest/manager.py:64]: What do you think of this approach instead?
```py
workflow_id = self._alias_dictionary.get(alias, workflow_id)
return self._workflows[workflow_id]
```
You can follow a similar pattern in the other methods, then delete `_check_workflow_alias`

File: src/quest/manager.py (hunk 6 above)
Comment (cwebbles) [reply src/quest/manager.py:170]: How can I protected these and also call them in the manager_wrappers?

**Pull request merged**